src/v/storage/tests/compaction_fuzz_test.cc

```diff
@@ -0,0 +1,196 @@
+            builder.add_raw_kv(to_iobuf(key), to_iobuf(val));
+        }
+    }
+    return std::move(builder).build();
+}
+
+static fragmented_vector<model::record_batch>
+generate_random_record_batches(int num, int cardinality) {
+    fragmented_vector<model::record_batch> result;
+    std::vector<std::optional<ss::sstring>> keys;
+    std::vector<std::optional<ss::sstring>> values;
+    std::vector<model::record_batch_type> types{
+      model::record_batch_type::raft_configuration,
+      model::record_batch_type::raft_data,
+      model::record_batch_type::archival_metadata,
+    };
+    for (int i = 0; i < cardinality; i++) {
+        if (i == 0) {
+            keys.emplace_back(std::nullopt);
+        } else {
+            keys.emplace_back(
+              random_generators::gen_alphanum_string(20, false));
+        }
+        values.emplace_back(random_generators::gen_alphanum_string(20, false));
+    }
+    // Generate actual batches
+    model::offset current{0};
+    for (int i = 0; i < num; i++) {
+        result.emplace_back(make_random_batch(
+          current,
+          false,
+          random_generators::random_choice(types),
+          keys,
+          values,
+          random_generators::get_int(1, 10)));
+        current = model::next_offset(result.back().last_offset());
+    }
+    return result;
+}
+
+/// Offset translator state observed at some point in time
+struct ot_state {
+    std::deque<model::offset> gap_offset;
+    std::deque<int64_t> gap_length;
+};
+
+/// Consumer that builds the map of all non-data
+/// batches!
+struct ot_state_consumer {
+    ss::future<ss::stop_iteration> operator()(model::record_batch rb) {
+        if (rb.header().type != model::record_batch_type::raft_data) {
+            // save information about the non-data batch
+            st->gap_offset.push_back(rb.base_offset());
+            st->gap_length.push_back(rb.record_count());
+        }
+        co_return ss::stop_iteration::no;
+    }
+
+    void end_of_stream() {}
+
+    ot_state* st;
+};
+
+/// Insert data into the log and maintain particular
+/// segment arrangement. The arrangement is defined
+/// by the set of segment base offset values.
+ss::future<ot_state> arrange_and_compact(
+  const fragmented_vector<model::record_batch>& batches,
+  std::deque<model::offset> arrangement) {
+    std::sort(arrangement.begin(), arrangement.end());
+    storage::disk_log_builder b1;
+    std::exception_ptr error = nullptr;
+    co_await b1.start();
+    try {
+        for (const auto& b : batches) {
+            co_await b1.add_batch(b.copy());
+            if (
+              !arrangement.empty() && b.base_offset() >= arrangement.front()) {
+                arrangement.pop_front();
+                co_await b1.get_disk_log_impl().force_roll(
+                  ss::default_priority_class());
+            }
+        }
+        ss::abort_source as;
+        co_await b1.apply_compaction(storage::compaction_config(
```

Comment:
This only applies adjacent segment compaction.

Should we extend this to test sliding window compaction?

Reply:
yes, I think so